### PR TITLE
Operations WIP

### PR DESCRIPTION
**Pull request by mattcollier**
- [x] Blocks no longer contain references to events
- [x] Events contain references to blockHeight
- [ ] Store operations separately from events using the same pattern as blocks/events 

Does order of events in the block need to be preserved?

### CODE REVIEW

File: lib/ledgerBlockStorage.js [hunk 1]
```diff
@@ -26,55 +26,58 @@ module.exports = class LedgerBlockStorage {
 
   /**
    * Adds a block in the ledger given a block, metadata associated with the
-   * block, and a set of options.
+   * block
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: Needs a period `.`

File: lib/ledgerBlockStorage.js [hunk 2]
```diff
@@ -26,55 +26,58 @@ module.exports = class LedgerBlockStorage {
...
-    if(typeof options === 'function') {
-      callback = options;
-      options = {};
-    }
+  add({block, meta}, callback) {
     if(!meta.blockHash) {
       return callback(new BedrockError(
         'A block hash for the given block was not specified.',
-        'BadRequest', {meta}));
+        'DataError', {meta}));
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: `DataError` seems vague. Seems like we should be throwing a `TypeError` for misuse of the API (or passing it to the callback). Thoughts?

File: lib/ledgerBlockStorage.js [hunk 3]
```diff
@@ -26,55 +26,58 @@ module.exports = class LedgerBlockStorage {
...
-      }
+        this.collection.insert(record, database.writeOptions, (err, result) => {
+          if(err) {
+            return callback(err);
+          }
+          callback(null, result.ops[0]);
+        });
+      },
+      update: ['insert', (results, callback) => {
+
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: Remove extra whitespace

File: lib/ledgerBlockStorage.js [hunk 4]
```diff
@@ -26,55 +26,58 @@ module.exports = class LedgerBlockStorage {
...
+        const query = {eventHash: {$in: event.map(h => database.hash(h))}};
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: Hmm, unfortunately we know that a `query` is itself a MongoDB document... so we can go over the limit here with this query. We'd need to batch the update. Can leave it as a FIXME and an issue if we want to skip this for now and pick it up in the next few weeks instead.

File: lib/ledgerBlockStorage.js [hunk 5]
```diff
@@ -26,55 +26,58 @@ module.exports = class LedgerBlockStorage {
...
+        this.eventCollection.update(
+          query, {$set: {'meta.blockHeight': blockHeight}},
+          database.writeOptions, callback);
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: What happens if we crash in the middle of this call? How do we recover and make sure all of the events are updated? Does the consensus algorithm handle that properly? I think the current code (at least in continuity) assumes writing the block always works ... so we'd need to be able to "finish adding" a block or something.

File: lib/ledgerBlockStorage.js [hunk 6]
```diff
@@ -641,36 +644,10 @@ module.exports = class LedgerBlockStorage {
...
-    const query = {
-      eventHash: {
-        $in: hashes
-      }
-    };
-    const projection = {
-      event: 1,
-      'meta.eventHash': 1,
-      _id: 0
-    };
```
Comment (dlongley) [inline lib/ledgerBlockStorage.js]: Yay for getting rid of that!

File: lib/ledgerEventStorage.js [hunk 7]
```diff
@@ -199,6 +197,9 @@ module.exports = class LedgerEventStorage {
     this.collection.find(query).count(callback);
   }
 
+  // FIXME: this seems like a very expensive operation and I don't think
+  // this API is in use anywhere, should it be removed?
```
Comment (dlongley) [inline lib/ledgerEventStorage.js]: Good FIXME ... I suspect we could get rid of it. Or at least make it private and not expose any docs on it until we know we need it.

File: lib/ledgerEventStorage.js [hunk 8]
```diff
@@ -231,16 +232,17 @@ module.exports = class LedgerEventStorage {
     if(typeof options.consensus === 'boolean') {
       query['meta.consensus'] = {$exists: options.consensus};
     }
+    // FIXME: is sort needed?  If so, an index on meta.eventHash is needed
     const queryOpts = {
-      sort: {eventHash: 1}
+      sort: {'meta.eventHash': 1}
```
Comment (dlongley) [inline lib/ledgerEventStorage.js]: I don't think we need a sort here? Not sure what it's for.

Comment (dlongley) [reply lib/ledgerEventStorage.js]: Do we still use this method at all?

**Comment by dlongley** [comment]
@mattcollier,

> Does order of events in the block need to be preserved?

Yeah, it does. So we'll need another property in the events for that.

**Comment by mattcollier** [comment]
@dlongley is the order if `event` topological order, or something else?

File: lib/ledgerBlockStorage.js (hunk 4 above)
Comment (mattcollier) [reply lib/ledgerBlockStorage.js]: @dlongley I will add a FIXME, keep in mind that the key size has gone from a 57 byte string to a 4 byte (32 bit) integer, a 92% size reduction!  So, where we were able to reference about ~300K events in a query, we can now reference ~4.2M.  So this seems well beyond the practical limits of what can/should go into a block, so now I'm thinking maybe we don't need a FIXME, what say you?

File: lib/ledgerBlockStorage.js (hunk 2 above)
Comment (mattcollier) [reply lib/ledgerBlockStorage.js]: Yes, I think a TypeError is appropriate and I will update all the other APIs as well.

**Comment by dlongley** [comment]
> is the order if event topological order, or something else?

It's topological + lexicographical to break ties.

File: lib/ledgerBlockStorage.js (hunk 4 above)
Comment (dlongley) [reply lib/ledgerBlockStorage.js]: Oh, yeah, we don't need a FIXME then.

File: lib/ledgerOperationStorage.js [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/*!
+ * Ledger block storage class.
```
Comment (dlongley) [inline lib/ledgerOperationStorage.js]: operation

**Pull request merged**